devel/query.lua

```diff
@@ -1,254 +1,558 @@
+possibly C++.
+
+This script takes your data selection eg.{table,unit,item,tile,etc.} then
+recursively iterates through it, outputting names and values of what it finds.
+
+As it iterates you can have it do other things, like search for a specific
+structure pattern (see lua patterns) or set the value of fields matching the
+selection and any search pattern specified.
+
+.. Note::
+
+    This is a recursive search function. The data structures are also recursive.
+    So there are a few things that must be considered (in order):
+
+        - Is the search depth too high? (Default: 7)
+        - Is the data capable of being iterated, or does it only have a value?
+        - How can the data be iterated?
+        - Is the iteration count for the data too high? (Default: 257)
+        - Does the user want to exclude the data's type?
+        - Is the data recursively indexing (eg. A.B.C.A.*)?
+        - Does the data match the search pattern?
+
+.. Warning::
+
+  This is a recursive script that's primary use is to search recursive data
+  structures. You can, fairly easily, cause an infinite loop. You can even
+  more easily run a query that simply requires an inordinate amount of time
+  to complete.
+
+.. Tip::
+
+  Should the need arise, you can kill the command from another shell.
+  Just run ``kill-lua`` from another terminal with the help of ``dfhack-run``.
+  ``$ dfhack-run kill-lua``
+
+Usage examples::
+
   devel/query -unit -getfield id
+  devel/query -unit -search STRENGTH
+  devel/query -unit -search physical_attrs -maxdepth 2
+  devel/query -tile -search dig
+  devel/query -tile -search "occup.*carv"
+  devel/query -table df -maxdepth 2
+  devel/query -table df -maxdepth 2 -excludekinds s -excludetypes fsu -oneline
+  devel/query -table df.profession -findvalue FISH
+  devel/query -table df.global.ui.main -maxdepth 0
+  devel/query -table df.global.ui.main -maxdepth 0 -oneline
+  devel/query -table df.global.ui.main -maxdepth 0 -1
 
 **Selection options:**
 
-These options are used to specify where the query will run,
-or specifically what key to print inside a unit.
+``-tile``
+  Selects the highlighted tile's block, and then
+  uses the tile's local position to index the 2D data.
+
+``-block``
+  Selects the highlighted tile's block.
+
+``-unit``
+  Selects the highlighted unit
 
-``-unit``:              Selects the highlighted unit
+``-item``
+  Selects the highlighted item.
 
-``-item``:              Selects the highlighted item.
+``-plant``
+  Selects the highlighted plant.
 
-``-tile``:              Selects the highlighted tile's block and then attempts to find the tile, and perform your queries on it.
+``-building``
+  Selects the highlighted building.
 
-``-table <value>``:     Selects the specified table (ie. 'value').
+``-job``
+  Selects the highlighted job.
 
-                        Must use dot notation to denote sub-tables.
-                        (eg. ``-table df.global.world``)
+``-script``
```

Comment:
Looks like this should be `-module`

Reply:
Also, this option _only_ works for scripts, not modules, so I think it should likely be renamed back to `-script`

Reply:
Hmm I thought dorf_tables was a module.

Reply:
I guess I can see how it makes sense either way. However, I was reading "script/module" as "a script that can be included with reqscript() *or* a library that can be included with require()". The latter is not supported but is arguably also a module. (I would tend to define a module as a file that can be included. This includes some scripts, like dorf_tables, but not all scripts, and also includes other types of files.)

Reply:
Honestly I'm a bit confused still cause of the ```--@ module = true``` at the top of the file. I guess it's only needed for allowing the file to be imported and doesn't really mean the file is a module.. but maybe somewhere between

Reply:
Sure, you could say that `--@ module=true` makes a script a module. What I am saying is that there are _other things_ that are also "modules", such as anything you include with `require()`. The flag in question only supports modules that are scripts, not all types of modules.